### PR TITLE
[x86] Don't pop arguments when returning from a managed-to-native wrapper.

### PR DESCRIPTION
**Pull request by madewokherd**
Callers do not expect these methods to pop arguments, resulting in a
wrong stack pointer.

For xamarin bug 30146.

### CODE REVIEW

File: mono/mini/mini-x86.c [hunk 1]
```diff
@@ -5425,7 +5425,8 @@ mono_arch_emit_epilog (MonoCompile *cfg)
 	if (need_stack_frame)
 		x86_leave (code);
 
-	if (CALLCONV_IS_STDCALL (sig)) {
+	if (CALLCONV_IS_STDCALL (sig) &&
+		method->wrapper_type != MONO_WRAPPER_MANAGED_TO_NATIVE) {
 		MonoJitArgumentInfo *arg_info = alloca (sizeof (MonoJitArgumentInfo) * (sig->param_count + 1));
```
Comment (vargaz) [inline mono/mini/mini-x86.c]: Why is this needed ? The managed-to-native wrappers are normal managed methods, they shouldn't have an stdcall calling convention.

Comment (madewokherd) [inline mono/mini/mini-x86.c]: Hm, currently we don't change call_convention between the wrapper method and the eventual native call (in mono_marshal_emit_native_wrapper), so for a stdcall function, they must both be stdcall. However, we do set the signature's pinvoke field, and we check that in several other places. So perhaps the real problem is that CALLCONV_IS_STDCALL can be true for a non-pinvoke signature. Does that make sense to you?

**Comment by vargaz** [comment]
The managed-to-wrapper method needs to use the default calling convention, that seems to be the problem.

**Comment by vargaz** [comment]
Fixing CALLCONV_IS_STDCALL should be enough, I don't think we ever want to call non-pinvoke methods with stdcall.

**Comment by madewokherd** [comment]
Makes sense to me, I updated the branch to fix the macro instead.

**Comment by madewokherd** [comment]
Whoops, sig->pinvoke needs more parens, don't merge this version.
